checkpoint/orbax/checkpoint/_src/path/deleter.py

```diff
@@ -67,14 +68,54 @@ def __init__(
       directory: refer to CheckpointManager.directory
       todelete_subdir: refer to CheckpointManagerOptions.todelete_subdir
       name_format: refer to CheckpointManager._name_format
+      is_hns_enabled: refer to CheckpointManagerOptions.is_hns_enabled
       duration_metric: the name of the total delete duration metric
     """
     self._primary_host = primary_host
     self._directory = directory
     self._todelete_subdir = todelete_subdir
     self._name_format = name_format
+    self._is_hns_enabled = is_hns_enabled
     self._duration_metric = duration_metric
 
+  def _list_dirs_recursively(self, path: epath.Path):
+    """Lists all directories recursively under the given path."""
+    if path.is_dir():
+      yield path
+    for p in path.iterdir():
+      if p.is_dir():
+        yield from self._list_dirs_recursively(p)
+
+  def _rmtree(self, path: epath.Path):
+    """Recursively deletes a path.
+
+    For a hierarchical namespace bucket, `path.rmtree()` only removes objects,
+    leaving all the empty parent folders intact. Here we manually delete the
+    empty folders recursively.
+
+    Args:
+      path: the path to delete.
+    """
+    # Step 1: Delete all files within the tree.
+    path.rmtree()
+
+    # Step 2: For HNS, clean up the remaining empty directory structure.
+    if self._is_hns_enabled:
+      # We need to delete from leaf to root.
+      # 1. Find all subdirectories within the path.
+      # 2. Sort them by depth (deepest first).
+      # 3. Delete each one.
+
+      # find all sub-directories.
+      all_dirs = [p for p in self._list_dirs_recursively(path) if p.is_dir()]
+
+      # Sort by the number of path parts, in reverse.
+      # This ensures `.../a/b/c` is processed before `.../a/b`.
+      all_dirs.sort(key=lambda p: len(p.parts), reverse=True)
+
+      for directory in all_dirs:
+        directory.rmdir()
```

Comment:
I'm afraid this might not work as expected.

https://github.com/google/etils/blob/73fda15bd6136806ad1b5eff6214867dc47de412/etils/epath/gpath.py#L211-L225

`rmdir` will fallback to `rmtree` and we may hit the same issue with unrecognised HNS.